### ISSUE TITLE
Scaling Orchestration

### ISSUE THREAD

**Issue by meowveen** [lyft/flinkk8soperator#146]
Hi,

I have a question regarding scaling the parallelism value.

For now, we are doing the manual way, whereby we edit the FlinkApplication YAML to change the parallelism value, apply the YAML, and let the operator transition from the old job cluster to the newly created job cluster.

Is there any recommended way to automatically orchestrate the above, similar to how HPA works in K8S, whereby HPA will kick in once the metrics exceed a certain value?

**Comment by anandswaminathan** [comment]
I am not aware of something that is available (or straightforward) to achieve this. We have wanted to build some functionality like this in the Operator, we are yet to design it.

**Comment by jirislav** [comment]
Hi, we would also be interested into an `auto-scaling` feature of the operator, since we often need to solve the problem of too low resources, resulting in high back-pressure and a need for a manual action.

Do you have some design yet? Or would you prefer to collaborate on this design?

**Comment by JRemitz** [comment]
It seems this thread is out of date. Since May, 2021, the Flink team has introduced [Reactive Mode](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/deployment/elastic_scaling/#reactive-mode) with Flink `v1.13`.  This supports [horizontal pod autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) as long as it is supported within your Kubernetes cluster (metrics-server deployed as well).

If I'm thinking through this correctly, we already have access to the Flink configuration updates to enable reactive mode:
```yaml
scheduler-mode: reactive
```
 and by deploying a container with Flink `v1.13+`.  The only additional need would be to include HPA with the Operator when it creates the necessary resources, and any rbac associated with creating that.

I can test appending HPA on after the fact, I haven't yet. But this should be enough in theory to get this working unless there are other Operator considerations that need to be made for managing the deployment with HPA.

More info on [Reactive Mode](https://flink.apache.org/2021/05/06/reactive-mode.html).

**Comment by anandswaminathan** [comment]
@JRemitz Sounds good. I guess there is one scenario, the deployment object created by the operator will need min/max. no ?

**Comment by JRemitz** [comment]
Yes, really the entire HPA definition for CPU rules. I'm not overly familiar with the CRD structure. Can you reference existing API objects as complex structures for validation?
